crypto/modes/gcm128.c

```diff
@@ -720,6 +725,39 @@ void gcm_gmult_clmul_rv64i_zbb_zbc(u64 Xi[2], const u128 Htable[16]);
 # endif
 #endif
 
+/*
+ * Non-static non-optimized version possibly used by
+ * AES-GCM-SIV for POLYVAL
+ * Optimized versions are used based on logic above and in
+ * CRYPTO_gcm128_init()
+ */
+void ossl_gcm_init_4bit(u128 Htable[16], const u64 H[2])
+{
+#ifdef HAVE_GCM_INIT_4BIT
+    gcm_init_4bit(Htable, H);
+#endif
```

Comment:
Maybe I'm missing something? It seems like these functions will be no-ops in some conditions which is a problem.

Reply:
Yes, they would be no-ops, but only for cases where *other* (i.e. assembly) `gcm_init_4bit`, `gcm_gmult_4bit`, and `gcm_ghash_4bit` functions (with suffixes) would be used, and not these iterations.
This code `#ifdef`s these functions out if they're not defined (otherwise linking would fail) and subsequently not used.

The AES-GCM-SIV code uses architecture-specific assembly versions of these functions when they're available, and uses so using these functions doesn't matter. The logic for that selection is in `cipher_aes_gcm_siv_polyval.c`.

I admit it's ugly, but the GCM code in general is ugly:
* there's an `8bit` and a `1bit` GHASH/GMULT implementation that are never used
* there are assembly versions of these functions for some architectures, but not others (which use C implementations)
* In the case of x86, it doesn't define these functions in assembly, but instead, defines them with an `_x86` suffix
* and there's probably more!

I put the ghash/gmult/init function selection logic into the AES-GCM-SIV code because I didn't want to break anything in `gcm128.c`!

The cross-compiles and tests for the various architectures pass, so it's all coming together.

If desired, I'd offer to make an additional PR that cleans up gcm128.c:
* Remove the `1bit` implementations
* Remove the `8bit` implementations
* Consolidate the various architecture selection code to be used in both gcm128.c and with AES-GCM-SIV.

Reply:
If it is guaranteed that these symbols are not being called when they are no-ops, why not just not define the symbols at all in that case?

If there is some reason why the symbols must be defined even though it is known they are never called, I would add a call to `OPENSSL_assert(false)` or `OPENSSL_die`.

Reply:
It's terribly messy due to the static-but-not-always-defined nature of these functions. The logic for deciding which is in multiple places (again, because I didn't want to risk breaking `gcm128.c`). I could just leave out the tip commit which adds the `ossl_` functions and makes the `gcm` functions non-static. @t8m was ok with that.

Putting in the asserts is not useful if they would subsequently be removed in a follow-up PR that focuses on cleaning up `gcm128.c`.

While a clean up of `gcm128.c` could stand on its own, it could be part of this (but as a distinct commit).

Reply:
Yeah, just revert this part of the change. If you'd be willing to do the cleanup PR later it would be nice.

Reply:
@t8m agreed. 

It might be better to evaluate a cleanup of `gcm128.c` first... as it would change the ghash interface in this PR quite a bit, and I don't see a reason to commit code that is rewritten within a week...